src/client.rs

```diff
@@ -447,6 +464,15 @@ pub(crate) unsafe extern "C" fn native_error_cb<C: ClientContext>(
     context.error(error, reason.trim());
 }
 
+pub(crate) unsafe extern "C" fn native_oauth_refresh_cb<C: ClientContext>(
+    client: *mut RDKafka,
+    oauthbearer_config: *const i8,
+    opaque: *mut c_void,
+) {
+    let context = &mut *(opaque as *mut C);
+    context.refresh_oauth_token(client, oauthbearer_config);
```

Comment:
Should it return a boolean so we can then call `oauthbearer_set_token()` or 
`oauthbearer_set_token_failure()`?

Reply:
Hoping that I'm understanding what you want here...

It'd be cool to put the unsafe calls to those functions in this block, but I'm not sure it makes sense.

For instance, I don't think we should assume what the value of parameters like `md_lifetime_ms` should be, and I don't think we should complicate `ClientContext` so that the user has a way to provide that value through context by default.

Also, I don't really want to assume anything about how the user handles errors in this process. By moving the call to `oauthbearer_set_token` into this block, we make it so that the user has a lot of flexibility in handling errors in generating the token (since that's code that they are going to write), but they have no flexibility in handling errors that get reported back via the `errstr` value - they would be bound to whatever we decide to do in this block.

I can write a pseudo prototype of a callback function to make things a bit clearer if needed.

Reply:
The callback could return the information needed for us to call `oauthbearer_set_token` here. One option would be to return `Result<SetTokenData, Error>`. The user can still have some custom error handling logic in the callback, but we'll take care of enforcing the [contract](https://docs.confluent.io/platform/current/clients/librdkafka/html/classRdKafka_1_1OAuthBearerTokenRefreshCb.html) that we must call `oauthbearer_set_token_failure` or `oauthbearer_set_token` in the callback.

Alternatively, if we really want the user to have all the flexibility in handling errors, we can return `Option<SetTokenData>`.

In any case, should this PR also include bindings for `oauthbearer_set_token_failure` or `oauthbearer_set_token`?

Reply:
Take a look at the most recent commit and let me know what you think. The only thing that's missing right now is support for SASL extensions, for which I would be happy to add support in my free time after this PR.

Reply:
Looks great!